File: android/lib_shared/src/main/java/edu/hawhamburg/shared/scenegraph/CubeNode.java
/**
 * Diese Datei gehört zum Android/Java Framework zur Veranstaltung "Computergrafik für
 * Augmented Reality" von Prof. Dr. Philipp Jenke an der Hochschule für Angewandte
 * Wissenschaften (HAW) Hamburg. Weder Teile der Software noch das Framework als Ganzes dürfen
 * ohne die Einwilligung von Philipp Jenke außerhalb von Forschungs- und Lehrprojekten an der HAW
 * Hamburg verwendet werden.
 * <p>
 * This file is part of the Android/Java framework for the course "Computer graphics for augmented
 * reality" by Prof. Dr. Philipp Jenke at the University of Applied (UAS) Sciences Hamburg. Neither
 * parts of the framework nor the complete framework may be used outside of research or student
 * projects at the UAS Hamburg.
 */
package edu.hawhamburg.shared.scenegraph;

import android.opengl.GLES20;

import java.util.ArrayList;
import java.util.List;

import edu.hawhamburg.shared.math.AxisAlignedBoundingBox;
import edu.hawhamburg.shared.math.Matrix;
import edu.hawhamburg.shared.math.Vector;
import edu.hawhamburg.shared.rendering.RenderVertex;
import edu.hawhamburg.shared.rendering.Shader;
import edu.hawhamburg.shared.rendering.ShaderAttributes;
import edu.hawhamburg.shared.rendering.VertexBufferObject;

/**
 * Representation of a cuboid with different dimensions in x-, y- and
 * z-direction.
 *
 * @author Philipp Jenke
 */
public class CubeNode extends LeafNode {

    /**
     * Cube side length
     */
    private double sideLength;

    /**
     * VBO.
     */
    private VertexBufferObject vbo = new VertexBufferObject();

    /**
     * Constructor.
     */
    public CubeNode(double sideLength) {
        this.sideLength = sideLength;
        createVbo();
    }

    private void createVbo() {
        List<RenderVertex> renderVertices = new ArrayList<RenderVertex>();

        Vector p0 = new Vector(-sideLength, -sideLength, -sideLength);
        Vector p1 = new Vector(sideLength, -sideLength, -sideLength);
        Vector p2 = new Vector(sideLength, sideLength, -sideLength);
        Vector p3 = new Vector(-sideLength, sideLength, -sideLength);
        Vector p4 = new Vector(-sideLength, -sideLength, sideLength);
        Vector p5 = new Vector(sideLength, -sideLength, sideLength);
        Vector p6 = new Vector(sideLength, sideLength, sideLength);
        Vector p7 = new Vector(-sideLength, sideLength, sideLength);
        Vector n0 = new Vector(0, 0, -1);
        Vector n1 = new Vector(1, 0, 0);
        Vector n2 = new Vector(0, 0, 1);
        Vector n3 = new Vector(-1, 0, 0);
        Vector n4 = new Vector(0, 1, 0);
        Vector n5 = new Vector(0, -1, 0);
        Vector color = new Vector(0.25, 0.75, 0.25, 1);

        AddSideVertices(renderVertices, p0, p1, p2, p3, n0, color);
        AddSideVertices(renderVertices, p1, p5, p6, p2, n1, color);
        AddSideVertices(renderVertices, p4, p7, p6, p5, n2, color);
        AddSideVertices(renderVertices, p0, p3, p7, p4, n3, color);
        AddSideVertices(renderVertices, p2, p6, p7, p3, n4, color);
        AddSideVertices(renderVertices, p5, p1, p0, p4, n5, color);

        vbo.setup(renderVertices, GLES20.GL_TRIANGLES);
    }

    /**
     * Add 4 vertices to the array
     */
    private void AddSideVertices(List<RenderVertex> renderVertices, Vector p0,
                                 Vector p1, Vector p2, Vector p3, Vector normal, Vector color) {
        renderVertices.add(new RenderVertex(p3, normal, color));
        renderVertices.add(new RenderVertex(p2, normal, color));
        renderVertices.add(new RenderVertex(p0, normal, color));
        renderVertices.add(new RenderVertex(p2, normal, color));
        renderVertices.add(new RenderVertex(p1, normal, color));
        renderVertices.add(new RenderVertex(p0, normal, color));
    }

    @Override
    public void drawGL(RenderMode mode, Matrix modelMatrix) {
        ShaderAttributes.getInstance().setShaderModeParameter(Shader.ShaderMode.PHONG);
        if (mode == RenderMode.REGULAR) {
            vbo.draw();
        }
    }

    @Override
    public AxisAlignedBoundingBox getBoundingBox() {
        return new AxisAlignedBoundingBox(new Vector(-sideLength, -sideLength, -sideLength),
                new Vector(sideLength, sideLength, sideLength));
    }
}

File: cs/cg_cs/src/scenegraph/CubeNode.cs
﻿using OpenTK;
using OpenTK.Graphics;
using OpenTK.Graphics.OpenGL;
using System.Collections.Generic;

namespace computergraphics
{
	/** 
	 * Draws a cube.
	 * */
	public class CubeNode : LeafNode
	{

		/**
		 * Cube is centered at the origin and its sidelength is twice this value.
		 * */
		private float lX;
		private float lY;
		private float lZ;

		private VertexBufferObject vbo = new VertexBufferObject();

		public CubeNode(float sideLength) : this(sideLength, sideLength, sideLength)
		{
		}

		public CubeNode(float lX, float lY, float lZ) : this(lX, lY, lZ, new Vector3(0, 0, 0))
		{
		}

		public CubeNode(float lX, float lY, float lZ, Vector3 center)
		{
			this.lX = lX / 2.0f;
			this.lY = lY / 2.0f;
			this.lZ = lZ / 2.0f;
			CreateVBO(center);
		}

		public override void DrawGL(RenderMode mode, Matrix4 modelMatrix)
		{
			if (mode == RenderMode.REGULAR)
			{
				vbo.Draw();
			}
		}

		public override void TimerTick(int counter)
		{
		}

		void CreateVBO(Vector3 offset)
		{
			List<RenderVertex> renderVertices = new List<RenderVertex>();

			Vector3 p0 = Vector3.Add(new Vector3(-lX, -lY, -lZ), offset);
			Vector3 p1 = Vector3.Add(new Vector3(lX, -lY, -lZ), offset);
			Vector3 p2 = Vector3.Add(new Vector3(lX, lY, -lZ), offset);
			Vector3 p3 = Vector3.Add(new Vector3(-lX, lY, -lZ), offset);
			Vector3 p4 = Vector3.Add(new Vector3(-lX, -lY, lZ), offset);
			Vector3 p5 = Vector3.Add(new Vector3(lX, -lY, lZ), offset);
			Vector3 p6 = Vector3.Add(new Vector3(lX, lY, lZ), offset);
			Vector3 p7 = Vector3.Add(new Vector3(-lX, lY, lZ), offset);
			Vector3 n0 = new Vector3(0, 0, -1);
			Vector3 n1 = new Vector3(1, 0, 0);
			Vector3 n2 = new Vector3(0, 0, 1);
			Vector3 n3 = new Vector3(-1, 0, 0);
			Vector3 n4 = new Vector3(0, 1, 0);
			Vector3 n5 = new Vector3(0, -1, 0);

			Color4 color = Color4.Aquamarine;
			AddSideVertices(renderVertices, p0, p1, p2, p3, n0, color);
			AddSideVertices(renderVertices, p1, p5, p6, p2, n1, color);
			AddSideVertices(renderVertices, p4, p7, p6, p5, n2, color);
			AddSideVertices(renderVertices, p0, p3, p7, p4, n3, color);
			AddSideVertices(renderVertices, p2, p6, p7, p3, n4, color);
			AddSideVertices(renderVertices, p5, p1, p0, p4, n5, color);

			vbo.Setup(renderVertices, PrimitiveType.Quads);
		}

		/**
		 * Add 4 vertices to the render list.
		 * */
		private void AddSideVertices(List<RenderVertex> renderVertices, Vector3 p0, Vector3 p1, Vector3 p2, Vector3 p3, Vector3 normal, Color4 color)
		{
			renderVertices.Add(new RenderVertex(p3, normal, color));
			renderVertices.Add(new RenderVertex(p2, normal, color));
			renderVertices.Add(new RenderVertex(p1, normal, color));
			renderVertices.Add(new RenderVertex(p0, normal, color));
		}
	}
}

